src/lib/optimizer/join_graph_builder.cpp

```diff
@@ -0,0 +1,265 @@
+#include <algorithm>
+#include <numeric>
+#include <queue>
+#include <stack>
+
+#include "expression/expression_functional.hpp"
+#include "expression/lqp_column_expression.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "utils/assert.hpp"
+
+#include "join_graph_edge.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+std::optional<JoinGraph> JoinGraphBuilder::operator()(const std::shared_ptr<AbstractLQPNode>& lqp) {
+  // No need to create a join graph consisting of just one vertex and no predicates
+  if (_lqp_node_type_is_vertex(lqp->type)) return std::nullopt;
+
+  traverse(lqp);
+
+  /**
+   * Turn the predicates into JoinEdges and build the JoinGraph
+   */
+  auto edges = join_edges_from_predicates(_vertices, _predicates);
+  auto cross_edges = cross_edges_between_components(_vertices, edges);
+
+  edges.insert(edges.end(), cross_edges.begin(), cross_edges.end());
+
+  return JoinGraph{_vertices, edges};
+}
+
+void JoinGraphBuilder::traverse(const std::shared_ptr<AbstractLQPNode>& node) {
+  // Makes it possible to call traverse() on inputs without checking whether they exist first.
+  if (!node) return;
+
+  if (_lqp_node_type_is_vertex(node->type)) {
+    _vertices.emplace_back(node);
+    return;
+  }
+
+  switch (node->type) {
+    case LQPNodeType::Join: {
+      /**
+       * Cross joins are simply being traversed past. Outer joins are hard to address during join ordering and until we
+       * do outer joins are opaque: The outer join node is added as a vertex and traversal stops at this point.
+       */
+
+      const auto join_node = std::static_pointer_cast<JoinNode>(node);
+
+      if (join_node->join_mode == JoinMode::Inner) {
+        _predicates.emplace_back(join_node->join_predicate);
+      }
+
+      if (join_node->join_mode == JoinMode::Inner || join_node->join_mode == JoinMode::Cross) {
+        traverse(node->left_input());
+        traverse(node->right_input());
+      } else {
+        _vertices.emplace_back(node);
+      }
+    } break;
+
+    case LQPNodeType::Predicate: {
+      const auto predicate_node = std::static_pointer_cast<PredicateNode>(node);
+      _predicates.emplace_back(predicate_node->predicate);
+
+      traverse(node->left_input());
+    } break;
+
+    case LQPNodeType::Union: {
+      /**
+       * A UnionNode is the entry point to a disjunction, which is parsed starting from _parse_union(). Normal traversal
+       * is commenced from the node "below" the Union.
+       */
+
+      const auto union_node = std::static_pointer_cast<UnionNode>(node);
+
+      if (union_node->union_mode == UnionMode::Positions) {
+        const auto parse_result = _parse_union(union_node);
+
+        traverse(parse_result.base_node);
+        _predicates.emplace_back(parse_result.predicate);
+      } else {
+        _vertices.emplace_back(node);
+      }
+    } break;
```

Comment:
I think we usually have the break within the block

Reply:
I never do, and some other places don't do it either. I think having it outside the block makes it clearer that this break has nothing to do with the logic in the block and is just there since c++'s switch is a stone age tool